### PR TITLE
improve protected branch to add whitelist support

### PR DESCRIPTION
**Pull request by lunny**
This PR will add a protected branch edit page to enable or disable protected branch. It also add support to whitelist users or whitelist teams so that special users could push to the branch.

### CODE REVIEW

File: models/migrations/v39.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	switch {
+	case setting.UseSQLite3:
+		log.Warn("Unable to drop columns in SQLite")
+	case setting.UseMySQL, setting.UsePostgreSQL, setting.UseMSSQL, setting.UseTiDB:
+		if _, err := x.Exec("ALTER TABLE protected_branch DROP COLUMN can_push"); err != nil {
+			return fmt.Errorf("DROP COLUMN can_push: %v", err)
+		}
+	default:
```
Comment (lafriks) [inline models/migrations/v39.go]: Need also migration support for PostgreSQL and MSSQL

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/migrations/v39.go (hunk 1 above)
Comment (lunny) [reply models/migrations/v39.go]: I think I have done that

Comment (lafriks) [reply models/migrations/v39.go]: Sorry didn't see, my mistake :)

**Review by lafriks — APPROVED** [review]

File: routers/repo/setting_protected_branch.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		if err != nil {
+			ctx.Handle(500, "UpdateProtectBranch", err)
+			return
+		}
+		ctx.Flash.Success(ctx.Tr("repo.settings.update_protect_branch_success"))
+		ctx.Redirect(fmt.Sprintf("%s/settings/branches/%s", ctx.Repo.RepoLink, branch))
+	} else {
+		if protectBranch != nil {
+			if err := ctx.Repo.Repository.DeleteProtectedBranch(protectBranch.ID); err != nil {
+				ctx.Handle(500, "UpdateProtectBranch", err)
```
Comment (ethantkoenig) [inline routers/repo/setting_protected_branch.go]: Wrong error message

File: options/locale/locale_en-US.ini [hunk 3]
```diff
@@ -922,6 +922,16 @@ settings.protected_branch=Branch Protection
 settings.protected_branch_can_push=Allow push?
 settings.protected_branch_can_push_yes=You can push
 settings.protected_branch_can_push_no=You can not push
+settings.branch_protection = Branch Protection for <b>%s</b>
+settings.protect_this_branch = Protect this branch
+settings.protect_this_branch_desc = Disable force pushes and prevent from deletion.
```
Comment (ethantkoenig) [inline options/locale/locale_en-US.ini]: I would suggest `Disable force pushes and prevent deletion.`

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -922,6 +922,16 @@ settings.protected_branch=Branch Protection
...
+settings.update_protect_branch_success = Change branch protect options successfully.
```
Comment (ethantkoenig) [inline options/locale/locale_en-US.ini]: I would suggest `Branch protect options changed successfully`.

File: options/locale/locale_en-US.ini [hunk 5]
```diff
@@ -922,6 +922,16 @@ settings.protected_branch=Branch Protection
...
+settings.protect_whitelist_committers = Whitelist who can push to this branch
+settings.protect_whitelist_committers_desc = Add people or teams to whitelist of direct push to this branch. Users in whitelist will bypass require pull request check.
+settings.protect_whitelist_users = Users who can push to this branch
+settings.protect_whitelist_search_users = Search users
+settings.protect_whitelist_teams = Teams for which members of them can push to this branch
```
Comment (ethantkoenig) [inline options/locale/locale_en-US.ini]: I would suggest `Teams whose members can push to this branch`.

File: options/locale/locale_en-US.ini [hunk 6]
```diff
@@ -922,6 +922,16 @@ settings.protected_branch=Branch Protection
 settings.protected_branch_can_push=Allow push?
 settings.protected_branch_can_push_yes=You can push
 settings.protected_branch_can_push_no=You can not push
+settings.branch_protection = Branch Protection for <b>%s</b>
+settings.protect_this_branch = Protect this branch
+settings.protect_this_branch_desc = Disable force pushes and prevent from deletion.
+settings.update_protect_branch_success = Change branch protect options successfully.
+settings.protect_whitelist_committers = Whitelist who can push to this branch
+settings.protect_whitelist_committers_desc = Add people or teams to whitelist of direct push to this branch. Users in whitelist will bypass require pull request check.
```
Comment (ethantkoenig) [inline options/locale/locale_en-US.ini]: I would suggest `Add users or teams to this branch's whitelist. Whitelisted users bypass the typical push restrictions.`

File: models/migrations/v39.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+package migrations
+
+import (
+	"fmt"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"github.com/go-xorm/xorm"
```
Comment (ethantkoenig) [inline models/migrations/v39.go]: Blank space

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/migrations/v39.go (hunk 7 above)
Comment (lunny) [reply models/migrations/v39.go]: done.

File: options/locale/locale_en-US.ini (hunk 5 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: options/locale/locale_en-US.ini (hunk 6 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: options/locale/locale_en-US.ini (hunk 4 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: options/locale/locale_en-US.ini (hunk 3 above)
Comment (lunny) [reply options/locale/locale_en-US.ini]: done.

File: routers/repo/setting_protected_branch.go (hunk 2 above)
Comment (lunny) [reply routers/repo/setting_protected_branch.go]: done.

**Review by ethantkoenig — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM

File: models/branches.go [hunk 8]
```diff
@@ -57,14 +92,101 @@ func GetProtectedBranchBy(repoID int64, BranchName string) (*ProtectedBranch, er
...
+				protectBranch.WhitelistTeamIDs = append(protectBranch.WhitelistTeamIDs, teams[i].ID)
+			}
+		}
+	}
+
+	// Make sure protectBranch.ID is not 0 for whitelists
+	if protectBranch.ID == 0 {
+		if _, err = x.Insert(protectBranch); err != nil {
+			return fmt.Errorf("Insert: %v", err)
+		}
```
Comment (lunny) [inline models/branches.go:140]: should `return nil`

**Comment by bkcsoft** [comment]
@lunny From the screenshot, do we really need to differentiate between Users and Teams? Can't they be in the same list? (Gitea knows the difference internally anyhow)

**Comment by lunny** [comment]
@bkcsoft  For a non-organization repository, we only have users. For an organization repository, we will have users and teams. I think both are needed.

**Comment by bkcsoft** [comment]
@lunny Right, the list it returns would be different, but I still don't understand why it needs to be 2 lists?

**Comment by lunny** [comment]
Two lists is easy to implement. 😄  We can optimist it later. We still need pull request control, status check and approval control.

**Comment by lafriks** [comment]
I also would like to see there single list instead of two but I agree that it can be implemented later

File: models/org.go [hunk 9]
```diff
@@ -577,6 +577,11 @@ func (org *User) getUserTeamIDs(e Engine, userID int64) ([]int64, error) {
 		Find(&teamIDs)
 }
 
+// TeamsHaveAccessToRepo returns all teamsthat have given access level to the repository.
+func (org *User) TeamsHaveAccessToRepo(repoID int64, mode AccessMode) ([]*Team, error) {
```
Comment (ethantkoenig) [inline models/org.go]: I'd suggest renaming to `TeamsWithAccessToRepo`(and similarly `GetTeamsWithAccessToRepo`)

File: models/repo.go [hunk 10]
```diff
@@ -667,6 +667,42 @@ func (repo *Repository) CanEnableEditor() bool {
 	return !repo.IsMirror
 }
 
+// GetWriters returns all users that have write access to the repository.
+func (repo *Repository) GetWriters() (_ []*User, err error) {
+	return repo.getUsersWithAccesMode(x, AccessModeWrite)
+}
+
+// getUsersWithAccesMode returns users that have at least given access mode to the repository.
+func (repo *Repository) getUsersWithAccesMode(e Engine, mode AccessMode) (_ []*User, err error) {
```
Comment (ethantkoenig) [inline models/repo.go]: misspelled `Access`

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
will be part of #32

**Comment by lunny** [comment]
@ethantkoenig all done.

**Review by ethantkoenig — APPROVED** [review]

**Comment by LanaYuan** [comment]
my gitea version 1.2 the protect branch tool didn't work, but when i use version 1.1.3, it works well

**Comment by lunny** [comment]
@LanaYuan could you try on https://try.gitea.io

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
LGTM

**Pull request merged**